### ISSUE TITLE
Consider redesigning the framework engine 

### ISSUE THREAD

**Issue by OliverO2** [kotest/kotest#4293]
Currently, it is hard to introduce a simple change (like [adding a context element at the spec level](https://github.com/kotest/kotest/pull/4284#issuecomment-2286407856)) correctly. And it is hard to find out what causes a bug.

It seems like the entire engine structure could benefit from some serious refactoring.

Problems I'm seeing:

* At various places, functionality is plugged in, which produces effects at unrelated locations sometime later (a catch-all mechanism like an event bus).
* There are manager classes (runner, executor, executor delegate, listener, materializer) with unclear responsibilities.
* We have some ball of global-like state which orchestrates the execution, so there are many effects, which we cannot observe on the call stack.

I'm aware that some complexity is the result of delegating parts of the execution to other frameworks (e.g. in JS).

Yet, it would be cool if we could get better abstractions for a good DX (avoiding duplication, good mental model, clear responsibilities, good discoverability).

**Comment by sksamuel** [comment]
Do you have some concrete examples of what we could do to improve things?

I thought the the pipeline/interceptor approach was pretty clean personally. Maybe the spec runner/delegates complicates things?

And there is no real global state, the configuration object is passed down through the context (unless that's what you mean?)

**Comment by OliverO2** [comment]
> Do you have some concrete examples of what we could do to improve things?

I have a broad idea, like making everything a Spec can do local to a Spec. So a Spec should be capable of executing itself rather than some SpecExecutor interfering from outside.

I'm looking for an easier mental model: I know what a Spec is, and I know what a Test is. So I'd know instantly where to find/change/add functionality if it were only these. (Test containers seem to be a bit more vaguely defined.)

In contrast, in the current `Spec` case, with additional manager-like classes like `SpecExecutor` and `SpecExecutorDelegate`, I need to look inside who does what. I now have three class hierarchies to consider. `SpecExecutor` does not even exist on all platforms, while the `SpecExecutorDelegate` does.

Ideally, we could coalesce all of that into a `Spec` and have just one abstraction to deal with.

> I thought the the pipeline/interceptor approach was pretty clean personally. Maybe the spec runner/delegates complicates things?

 I think the pipeline/interceptor approach is excellent for providing extensibility for Kotest users. It is less ideal for configuring Kotest inside. We could have better cohesion, more explicit interdependencies and better discoverability by avoiding the loose coupling of interceptor chains.

Consider our case: If `beforeSpec` was not an `Extension` registered in a `TestConfiguration` but a direct method in a Spec, registering the callback in that Spec, it could easily provide the spec-level coroutine to that callback. And in the debugger, the stack would show the spec executing that callback as part of the spec's execution.

> And there is no real global state, the configuration object is passed down through the context (unless that's what you mean?)

Yes, that's what I meant with "global-like". The entire configuration is for every place to inspect. The same with `CoroutineContext.Elements`. Such stuff is definitely useful, but would better be used as a last resort, as we cannot easily find out where its effects unfold.

Does that make sense?

**Comment by OliverO2** [comment]
I've created a minimal example of what I'm imagining: https://github.com/OliverO2/kotlin-test-framework-prototype

**Comment by sksamuel** [comment]
Thanks

> > Do you have some concrete examples of what we could do to improve things?
> 
> I have a broad idea, like making everything a Spec can do local to a Spec. So a Spec should be capable of executing itself rather than some SpecExecutor interfering from outside.

I like the idea of knowing all spec stuff is in one place, but isn't that what we have at the moment? If you consider a Spec to be the users declaration of a spec, and then our SpecRunner stuff is the execution. It's good design imo to keep data and logic separate.

Let's say we did do it your way though, all the spec runner code would need to ultimately move into the spec class making it a monster of a class. I try to keep my stuff very small and isolated as I think it's easier to work with than 2000 line classes.

> I'm looking for an easier mental model: I know what a Spec is, and I know what a Test is. So I'd know instantly where to find/change/add functionality if it were only these. (Test containers seem to be a bit more vaguely defined.)
> 
> In contrast, in the current `Spec` case, with additional manager-like classes like `SpecExecutor` and `SpecExecutorDelegate`, I need to look inside who does what. I now have three class hierarchies to consider. `SpecExecutor` does not even exist on all platforms, while the `SpecExecutorDelegate` does.
> 
> Ideally, we could coalesce all of that into a `Spec` and have just one abstraction to deal with.

I agree that the spec delegate stuff is very complicated. That's partially because JS runs completely differently, plus we have three isolation modes.

> > I thought the the pipeline/interceptor approach was pretty clean personally. Maybe the spec runner/delegates complicates things?
> 
> I think the pipeline/interceptor approach is excellent for providing extensibility for Kotest users. It is less ideal for configuring Kotest inside. We could have better cohesion, more explicit interdependencies and better discoverability by avoiding the loose coupling of interceptor chains.

The pipeline approach is meant to make it easier for us to add new functionality to Kotest. You just add an interceptor and plug it in. I think the loose coupling makes things easier to test in isolation and easier to add new things than having everything in one place tightly coupled. If you want to add a new context element at the spec level, you just add a spec interceptor.

> Consider our case: If `beforeSpec` was not an `Extension` registered in a `TestConfiguration` but a direct method in a Spec, registering the callback in that Spec, it could easily provide the spec-level coroutine to that callback. And in the debugger, the stack would show the spec executing that callback as part of the spec's execution.

But there's multiple ways to add beforeSpec - project level extensions. Eg the Spring extension. So your spec still needs to know about all the other beforeSpec listeners dotted about, so it needs to be provided with those. Then there's those registered using annotations, so the spec needs to look those up and instantiate them. Is that any different to what the SpecExtensions class is doing?

> > And there is no real global state, the configuration object is passed down through the context (unless that's what you mean?)
> 
> Yes, that's what I meant with "global-like". The entire configuration is for every place to inspect. The same with `CoroutineContext.Elements`. Such stuff is definitely useful, but would better be used as a last resort, as we cannot easily find out where its effects unfold.

I don't really follow this bit.

> Does that make sense?

I think honestly there's two parts here. a) is all the spec delegater / runner stuff complicated - yes. It is, and if we can simplify it I'm all for it. PR away!. 

But b) I think the demo you made is simple because it doesn't do anything other than run tests. Once you add in all the million features, it'll just be as complicated as the pipeline stuff today. Failfast, spec ordering, all the callbacks needed to feed into junit or intellij, checking for tags, detecting ignored tests and not running callbacks, wiring in assertion modes, supporting multiple isolation modes, and so on. In fact most of the interceptors we have are just running callbacks or setting up context elements. 

Plus things are always simpler when you're the one who wrote it in the first person as its natural to you.

If we can make some small steps to improve things and just keep iterating, why don't we do that ?

**Comment by OliverO2** [comment]
Thanks for looking into it!

I might not have presented my pain points in the best possible way so far, and I think the best way forward is to produce some tangible evidence. But first, I'd like to reply.

My motivation: What I'm observing is that I'm basically struggling with providing good quality reviews on engine stuff. There are so many moving parts and I often feel like I'm failing to consider everything that matters. I'd also like to be more efficient in making non-local changes and debugging.

I need a better immediate and correct understanding of structure, in particular with stuff I'm unfamiliar with or haven't touched for quite a while.

> It's good design imo to keep data and logic separate.

Wouldn't that end up in data classes which are manipulated by outside entities?

I'm generally seeing the best results with stringent OOD, where classes strive for cohesion, encapsulation is a thing and inter-class coupling is reduced to a minimum. That means putting logic where the corresponding data is, if possible.

Example: Methods in a `Spec` handle this Spec and would not interfere with another Spec. I can be immediately confident about this. In contrast, a `SpecExecutor` could be designed to handle a single spec, but also all of them. I cannot assume one or the other immediately, I'd have to look inside.

> I try to keep my stuff very small and isolated as I think it's easier to work with than 2000 line classes.
 
I try to do the same, considering tradeoffs: Splitting a 2000 line class at arbitrary boundaries into four tightly coupled 500+ line classes would not help.

Side note: While I think that a general design philosophy can provide useful guidance, I'm reluctant to apply "universal" rules. Context matters and everything needs to be reasonably balanced towards reaching our goals.

> I think the loose coupling makes things easier to test in isolation and easier to add new things than having everything in one place tightly coupled.

Yes, that's the upside. The downside is that it makes understanding the whole picture more difficult. And I feel like this can lead to write-only code.
 
> If you want to add a new context element at the spec level, you just add a spec interceptor.

...unless you detect that this interceptor is [not actually present in some places where you'd expect it](https://github.com/kotest/kotest/pull/4284#issuecomment-2286407856).

> I don't really follow this bit.

I'll try again, with coroutine contexts: These are basically global-like state (coroutine-hierarchy-local, to be precise). They are for everyone in a coroutine hierarchy to inspect. So their effects could unfold anywhere, implicitly. Unlike explicit parameters, global-like state makes it hard to reason about what happens, particularly in a large code base. Used sparingly, they are definitely useful.

> But b) I think the demo you made is simple because it doesn't do anything other than run tests.

Sure, it doesn't cover what Kotest currently does.

I think the prototype's approach has one bigger challenge:
* Integration with existing test frameworks. The idea here is to have a platform-specific invocation mechanism (a future Gradle plugin, Amper, CLI?) calling into the test framework's single top-level `suspend` function, like the one provided by `Module` class.

On the other hand, the prototype provides more functionality than it might appear:
* Like in Kotest, detection of ignored tests and not running callbacks is there.
* Unlike Kotest, focusing works at all levels.
* Unlike Kotest, all `before`/`after` interceptors work independently at multiple levels of nesting
* We have `SpecGroup`s, which come with group-specific spec configuration. One use case is to have separate scheduling variants like with `@DoNotParallelize`.
* Coroutines are scoped naturally by default across the test hierarchy (so none of #3577, #3705, #3759).

Where I'd like to deviate:
* I'd like to avoid providing multiple equivalent solutions for the same use case, since that would increase maintenance efforts without real gains. I'd prefer to keep a framework as small and simple as possible.

> Once you add in all the million features, it'll just be as complicated as the pipeline stuff today.

I'd really like to find out more and provide some numbers comparing the prototype with the equivalent parts of Kotest. I think concrete evidence is better than trying to guess a design's efficiency in the abstract.

> If we can make some small steps to improve things and just keep iterating, why don't we do that ?

Definitely, lets do that! Keep in mind that my approach is nothing that I'd expect to work in the near future.

But wouldn't it make sense to also explore alternatives? Complex products grow over time and usually become ever more complex. I think there is a time to grow, and there is a time to shrink stuff to its essence, based on what has been learned along the way.

**Comment by sksamuel** [comment]
> My motivation: What I'm observing is that I'm basically struggling with providing good quality reviews on engine stuff. There are so many moving parts and I often feel like I'm failing to consider everything that matters. I'd also like to be more efficient in making non-local changes and debugging.

That might be just because you're not working in this codebase every day. I struggle to understand stuff that was added without my involvement. For example the timesource stuff I still don't really understand why its useful but I trust it is.

> Wouldn't that end up in data classes which are manipulated by outside entities?

Yes exactly. Don't mix state and logic. That's what FP is all about and I'm a FP person really not an OO person. I think OO leads to systems that are harder to reason about.
>
> I'm generally seeing the best results with stringent OOD, where classes strive for cohesion, encapsulation is a thing and inter-class coupling is reduced to a minimum. That means putting logic where the corresponding data is, if possible.

This is something we'll have to disagree on. I've not been a fan of OO systems since I moved from Java to Scala in 2011 and learnt a lot more about FP paradigms.

> Example: Methods in a `Spec` handle this Spec and would not interfere with another Spec. I can be immediately confident about this. In contrast, a `SpecExecutor` could be designed to handle a single spec, but also all of them. I cannot assume one or the other immediately, I'd have to look inside.

Yeah I'm not sure it's simpler to have all the engine runtime stuff also shoved into the Spec superclass. But we can look into it.

> Side note: While I think that a general design philosophy can provide useful guidance, I'm reluctant to apply "universal" rules. Context matters and everything needs to be reasonably balanced towards reaching our goals.

I agree.
 
> > I think the loose coupling makes things easier to test in isolation and easier to add new things than having everything in one place tightly coupled.
> 
> Yes, that's the upside. The downside is that it makes understanding the whole picture more difficult. And I feel like this can lead to write-only code.

I think the write-only code is when you have giant classes personally. The big picture bit is true regardless of where you put things I think.

> > If you want to add a new context element at the spec level, you just add a spec interceptor.
> 
> ...unless you detect that this interceptor is [not actually present in some places where you'd expect it](https://github.com/kotest/kotest/pull/4284#issuecomment-2286407856).
> 
> > I don't really follow this bit.
> 
> I'll try again, with coroutine contexts: These are basically global-like state (coroutine-hierarchy-local, to be precise). They are for everyone in a coroutine hierarchy to inspect. So their effects could unfold anywhere, implicitly. Unlike explicit parameters, global-like state makes it hard to reason about what happens, particularly in a large code base. Used sparingly, they are definitely useful.

So you're saying that shoving things into the coroutine context is opaque because its not clear who is putting them where? If so, I agree with you. It's better to be strongly typed. It also means its harder to extend (eg the property test stuff. I'm loathe to go building it directly into the engine).

> > If we can make some small steps to improve things and just keep iterating, why don't we do that ?
> 
> Definitely, lets do that! Keep in mind that my approach is nothing that I'd expect to work in the near future.
> 
> But wouldn't it make sense to also explore alternatives? Complex products grow over time and usually become ever more complex. I think there is a time to grow, and there is a time to shrink stuff to its essence, based on what has been learned along the way.

I think we can do both at the same time? Start making small changes to push things in the direction you like. I think a good example here is the work @aSemy has been doing on the build. Lots of iteration. The obvious contender to start with is simplifying down the spec executor stuff. Or squashing / simplifying / removing some of the interceptors. We have time before 6.0 needs to be released so we can make big changes if needed.

**Comment by sksamuel** [comment]
> * Unlike Kotest, focusing works at all levels.

By the way, the only reason that focus is done only at the top level is because if you focus a nested test, root tests that run before it don't know that it's focused. And that will lead to 2000 issues being opened about why it doesn't work. So I just said, focus is a top level thing only.

(Bang works at all levels).

**Comment by OliverO2** [comment]
> That might be just because you're not working in this codebase every day.

Exactly. The same thing can also happen to my future self with code I wrote.

> For example the timesource stuff I still don't really understand why its useful but I trust it is.

What are you missing? Wasn't the flakiness of testing real-time constructs evident? And the stability of those tests once we were running them on virtual time?

> I've not been a fan of OO systems since I moved from Java to Scala in 2011
 
That's interesting. Isn't it a contradiction that you're using classes extensively in your own libs?

Things I consider hallmarks of well-written software:
1. How useful is it (including effectiveness and efficiency)?
2. How quickly can I obtain a correct understanding of the system, from the big picture down to the details of interest?
3. How efficiently can it evolve?

I am convinced that the main value of good OO designs is that these help with (2.) and (3.), because they let us to build accurate mental models quicker. Our brains are wired to use the generalization/specialization concept to understand the world. And OO mirrors that nicely. But I'm aware that many folks use inheritance for technical reasons against accurate semantic modeling, making the designs hard to understand and difficult to change. Other folks then generally advise against the use of inheritance because it is so often abused.

> So you're saying that shoving things into the coroutine context is opaque because its not clear who is putting them where?

Yes. They might be around, or they might not. You don't know. Coping with this requires good discipline.

> I think we can do both at the same time?

Yes.

For now, I'm using my approach as a lab, a small and simple system to explore what things could look like. Some results might then be directly fed back into Kotest, and these don't have to be big changes.

**Comment by sksamuel** [comment]
> > For example the timesource stuff I still don't really understand why its useful but I trust it is.
> 
> What are you missing? Wasn't the flakiness of testing real-time constructs evident? And the stability of those tests once we were running them on virtual time?

Yes it was flaky for sure. I thought we could have just adjusted the timeouts to make them more tolerant of slower systems. It's all good stuff, just saying that when you write it, it's easy to understand. When you don't, it's complicated :)

> > I've not been a fan of OO systems since I moved from Java to Scala in 2011
> 
> That's interesting. Isn't it a contradiction that you're using classes extensively in your own libs?

Classes are not exclusive to OO. Encapsulation, inheritance, those are the kinds of things I tend to avoid in FP. I use classes as namespaces for related functions. Or I like composition, and will have classes implement an interface, which is isomorphic to higher order functions.

> Things I consider hallmarks of well-written software:
> 
> 1. How useful is it (including effectiveness and efficiency)?
> 2. How quickly can I obtain a correct understanding of the system, from the big picture down to the details of interest?
> 3. How efficiently can it evolve?
> 
> I am convinced that the main value of good OO designs is that these help with (2.) and (3.), because they let us to build accurate mental models quicker. Our brains are wired to use the generalization/specialization concept to understand the world. And OO mirrors that nicely. But I'm aware that many folks use inheritance for technical reasons against accurate semantic modeling, making the designs hard to understand and difficult to change. Other folks then generally advise against the use of inheritance because it is so often abused.
> 
> > So you're saying that shoving things into the coroutine context is opaque because its not clear who is putting them where?
> 
> Yes. They might be around, or they might not. You don't know. Coping with this requires good discipline.
> 
> > I think we can do both at the same time?
> 
> Yes.
> 
> For now, I'm using my approach as a lab, a small and simple system to explore what things could look like. Some results might then be directly fed back into Kotest, and these don't have to be big changes.

I think your critique really comes down to two things:
1. Shoving non-strongly typed params into coroutine context rather than some strongly typed state object that's passed around.
2. The spec delegator stuff is complicated. It is, but that's partially because the JS executors work in a different way (no listener, but returning a promise) as well as needing three isolation modes.

If we cleaned up the above, you'd probably be happy is my take on it.

**Comment by sksamuel** [comment]
This PR might be a small first step in the right direction: https://github.com/kotest/kotest/pull/4590

**Comment by sksamuel** [comment]
See https://github.com/kotest/kotest/pull/4626

**Comment by sksamuel** [comment]
Since you built your own test framework, I assume you don't mind if I close this :)

**Comment by OliverO2** [comment]
That's fine! Turned out that to prove my point I actually had to draw [the rest of the owl](https://casnocha.com/2010/11/how-to-draw-an-owl.html). Hope there's some value in having done so.
